src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs

```diff
@@ -289,108 +295,30 @@ public virtual PipeTransmissionMode ReadMode
         // ---- PAL layer ends here ----
         // -----------------------------
 
-        private class ReadWriteAsyncParams
-        {
-            public ReadWriteAsyncParams() { }
-            public ReadWriteAsyncParams(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
-            {
-                this.Buffer = buffer;
-                this.Offset = offset;
-                this.Count = count;
-                this.CancellationHelper = cancellationToken.CanBeCanceled ? new IOCancellationHelper(cancellationToken) : null;
-            }
-            public byte[] Buffer { get; set; }
-            public int Offset { get; set; }
-            public int Count { get; set; }
-            public IOCancellationHelper CancellationHelper { get; private set; }
-        }
-
-        [SecurityCritical]
-        private unsafe static readonly IOCompletionCallback s_IOCallback = new IOCompletionCallback(PipeStream.AsyncPSCallback);
-
         [SecurityCritical]
-        private IAsyncResult BeginWrite(AsyncCallback callback, Object state)
-        {
-            ReadWriteAsyncParams readWriteParams = state as ReadWriteAsyncParams;
-            Debug.Assert(readWriteParams != null);
-            byte[] buffer = readWriteParams.Buffer;
-            int offset = readWriteParams.Offset;
-            int count = readWriteParams.Count;
-
-            if (buffer == null)
-            {
-                throw new ArgumentNullException("buffer", SR.ArgumentNull_Buffer);
-            }
-            if (offset < 0)
-            {
-                throw new ArgumentOutOfRangeException("offset", SR.ArgumentOutOfRange_NeedNonNegNum);
-            }
-            if (count < 0)
-            {
-                throw new ArgumentOutOfRangeException("count", SR.ArgumentOutOfRange_NeedNonNegNum);
-            }
-            if (buffer.Length - offset < count)
-            {
-                throw new ArgumentException(SR.Argument_InvalidOffLen);
-            }
-            if (!CanWrite)
-            {
-                throw __Error.GetWriteNotSupported();
-            }
-            CheckWriteOperations();
-
-            if (!_isAsync)
-            {
-                return _streamAsyncHelper.BeginWrite(buffer, offset, count, callback, state);
-            }
-            else
-            {
-                return BeginWriteCore(buffer, offset, count, callback, state);
-            }
-        }
-
-        [SecurityCritical]
-        unsafe private PipeStreamAsyncResult BeginWriteCore(byte[] buffer, int offset, int count,
-                AsyncCallback callback, Object state)
+        private unsafe Task WriteAsyncCorePrivate(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```

Comment:
Does the whole method need to allow unsafe, or could we scope it just to the line or two that needs it?